### PR TITLE
Nested input types with the same not supported

### PR DESCRIPTION
**Pull request by SimonCropp**
seems input types are keyed off the name, which is not unique. I assume the same would happen for input types in different namespaces

### CODE REVIEW

File: src/GraphQL.Tests/Bugs/Bug_nested_type_names_dont_conflict.cs [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+        }
+        public class MyInputClass2
+        {
+            public string Id { get; set; }
+
+            public class MyInput : InputObjectGraphType
+            {
+                public MyInput()
+                {
+                    Field<NonNullGraphType<StringGraphType>>("id");
```
Comment (Shane32) [inline src/GraphQL.Tests/Bugs/Bug_nested_type_names_dont_conflict.cs:66]:
```suggestion
                    Name = "MyInput2";
                    Field<NonNullGraphType<StringGraphType>>("id");
```
Is this not satisfactory?

Comment (SimonCropp) [reply src/GraphQL.Tests/Bugs/Bug_nested_type_names_dont_conflict.cs:66]: shouldnt this be supported OOTB?

Comment (sungam3r) [reply src/GraphQL.Tests/Bugs/Bug_nested_type_names_dont_conflict.cs:66]: OK

Comment (sungam3r) [reply src/GraphQL.Tests/Bugs/Bug_nested_type_names_dont_conflict.cs:66]: In general, the problem is wider and inherent to all grapj types. By default, the type name is defined in `GraphType.cs` as:
```
  private string GetDefaultName()
        {
            var type = GetType();

            var attr = type.GetCustomAttribute<GraphQLMetadataAttribute>();

            if (!string.IsNullOrEmpty(attr?.Name))
            {
                return attr.Name;
            }

            var name = type.Name.Replace('`', '_');
            if (name.EndsWith(nameof(GraphType), StringComparison.InvariantCulture))
                name = name.Substring(0, name.Length - nameof(GraphType).Length);

            return name;
        }
```

That is, only the name type is used without namespace and without parent types. This is done intentionally for ease of names. We _can_ change the algorithm to take into account paren types and namespaces, but it will be a breaking change. I suggest to use `GlobalSwitches`.

Comment (sungam3r) [reply src/GraphQL.Tests/Bugs/Bug_nested_type_names_dont_conflict.cs:66]: I'll post a PR.

**Comment by sungam3r** [comment]
@SimonCropp Thanks for the repro. Closing this as issue now is tracked in #2459 .

**Comment by Shane32** [comment]
I'm not exactly sure how this is a bug.  The default name of a type (including input object graph types) is derived from the class name.  There are other ways to create an duplicate implicit class name.  When a duplicate occurs, you should manually set the name of one or the other of the classes to a specified name, either via code or an attribute.  Besides the sample shown in the PR (putting the classes in different namespaces or as subclasses of different classes), here are four methods to create a graph type with the same name:

```cs
public class Widget : ObjectGraphType { }  // Name defaults to "Widget"

public class WidgetGraphType : ObjectGraphType { } // Name defaults to "Widget"

public class MyGraphType : ObjectGraphType
{
    public MyGraphType() {
        Name = "Widget";
    }
}

[GraphQLMetadata("Widget")]
public class MyGraphType2 : ObjectGraphType { }
```

Further, even if we wanted to "fix" it, the name is set before the schema is constructed.  You'd have to generate a longer name (e.g. from `FullName`), which doesn't make sense to me, since the graph type's constructor has no way of knowing what other types might or might not be added to the schema.

@sungam3r 's solution in #2459 does not address namespace differences.  However, at least it ONLY affect nested types.  But let's consider the consequences: For the above sample, now the input graph type names are `MyInputClass_MyInput` and `MyInputClass2_MyInput`.  The type names are necessary when used as a variable, so you'd be writing:

```gql
mutation ($arg: MyInputClass_MyInput) {
   doSomeStuff(arg: $arg)
}
```

It seems to me if someone runs into a duplicate type name, they should define the name they want, not rely on a global switch to make longer names.

In short, the only time I see this being either desirable or useful would be for writing tests.  And I don't think we should be adding this option for tests.

**Comment by sungam3r** [comment]
In general I agree. I thought it would not hurt to give a global option. In the end, the default behavior does not change. Perhaps someone will come in handy behavior.
